File: Graph/Graph_Breadth_First/GraphBFS.java
import java.util.Queue;
import java.util.LinkedList;

/**
 * Breadth-First Search implementation class.
 * @author Yasin Radi
 */
public class GraphBFS {

    // Num of vertices
    private int numVertex;

    // Adjacency list 
    private LinkedList<Integer>[] adj;

    // Constructor
    public GraphBFS(int n) {
        this.numVertex = n;
        this.adj = (LinkedList<Integer>[]) new LinkedList<?>[n];
        for(int i = 0; i < n; i++)
            this.adj[i] = new LinkedList<Integer>();
    };

    /**
     * Adds a new edge to the Graph.
     * @param nodeIndex     int
     * @param newNodeValue  int
     */
    public void addEdge(int nodeIndex, int newNodeValue) {
        this.adj[nodeIndex].add(newNodeValue);
    };

    /**
     * Performs a Breadth-First traversal to display
     * all nodes contained in Graph.
     * @param nodeIndex     int
    */
    public void BFS(int nodeIndex) {
        // Mark all the vertices as not visited
        boolean[] visited = new boolean[this.numVertex];
        for(int i = 0; i < this.numVertex; i++) 
            visited[i] = false;

        // Create a queue for BFS
        Queue<Integer> q = new LinkedList<>();

        // Mark the current node as visited and push it
        // to the queue
        visited[nodeIndex] = true;
        q.add(nodeIndex);

        // Builder to print result
        StringBuilder sb = new StringBuilder();

        while(!q.isEmpty()) {
            // Dequeue a vertex from the queue and print it
            int nodes = q.poll();
            sb.append(nodes).append(" ");

            // Get all adjacent vertices of the dequeued vertex.
            // If and adjacent has not been visited, then mark it
            // as visited and enqueue it.
            for(int element : this.adj[nodes]) {
                if(!visited[element]) {
                    visited[element] = true;
                    q.add(element);
                };
            };
        };

        System.out.println(sb.toString());
    };
};

File: Graph/Graph_Breadth_First/GraphBFS.cs
using System;
using System.Text;
using System.Collections.Generic;


/// <summary>
/// Breadth-First Search implementation class.
/// </summary>
public class GraphBFS
{

    // Num of vertices
    private int numVertex;

    // Adjacency list
    private LinkedList<int>[] adj;

    // Constructor
    public GraphBFS(int n) {
        this.numVertex = n;
        this.adj = new LinkedList<int>[n];
        for (int i = 0; i < n; i++)
            this.adj[i] = new LinkedList<int>();
    }

    /// <summary>
    /// Adds an edge to the graph.
    /// </summary>
    /// <param name="nodeIndex"> Index of node to add.</param>
    /// <param name="newNodeValue"> Value of the node being added.
    public void AddEdge(int nodeIndex, int newNodeValue) {
        adj[nodeIndex].AddLast(newNodeValue);
    }

    /// <summary>
    /// Performs a Breadth-First traversal to display
    /// all nodes contained in Graph.
    /// </summary>
    /// <param name="nodeIndex"> Index of node to start the BFS.</param>
    public void BFS(int nodeIndex) {
        // Mark all the vertices as not visited
        bool[] visited = new bool[numVertex];
        for (int i = 0; i < numVertex; i++)
            visited[i] = false;
        
        // Create a queue for BFS
        Queue<int> q = new Queue<int>();

        // Mark current node as visited and push it
        // to the queue
        visited[nodeIndex] = true;
        q.Enqueue(nodeIndex);

        // Builder to print result
        StringBuilder sb = new StringBuilder();

        while (q.Count > 0) {
            // Dequeues a vertex from the queue and prints it
            int node = q.Dequeue();
            sb.Append(node).Append(" ");

            /*
                Get all adjacent vertices of the dequeued vertex.
                If an adjacent has not been visited, mark it as 
                visited and enqueue it.
             */
             foreach (int element in adj[node]) {
                 if (!visited[element]) {
                     visited[element] = true;
                     q.Enqueue(element);
                 }
             }
        }

        Console.WriteLine(sb.ToString());
    }
}